### PR TITLE
When writing class elements show completion with allowed keywords and inheritted properties

### PR DESCRIPTION
**Pull request by sheetalkamat**
Handles #7158

### CODE REVIEW

File: src/harness/fourslash.ts [hunk 1]
```diff
@@ -3417,6 +3417,17 @@ namespace FourSlashInterface {
...
+        public allowedClassElementKeywords = [
+            "public",
+            "private",
+            "protected",
+            "static",
+            "abstract",
+            "readonly",
+            "get",
+            "set",
+            "constructor"
```
Comment (mhegazy) [inline src/harness/fourslash.ts]: also `async`

File: src/services/completions.ts [hunk 2]
```diff
@@ -1306,6 +1420,28 @@ namespace ts.Completions {
...
+        switch (kind) {
+            case SyntaxKind.PublicKeyword:
+            case SyntaxKind.ProtectedKeyword:
+            case SyntaxKind.PrivateKeyword:
+            case SyntaxKind.AbstractKeyword:
+            case SyntaxKind.StaticKeyword:
+            case SyntaxKind.ConstructorKeyword:
+            case SyntaxKind.ReadonlyKeyword:
+            case SyntaxKind.GetKeyword:
+            case SyntaxKind.SetKeyword:
```
Comment (mhegazy) [inline src/services/completions.ts]: async

File: src/services/completions.ts [hunk 3]
```diff
@@ -953,6 +1009,49 @@ namespace ts.Completions {
...
+        }
+
+        /**
+         * Returns the immediate owning class declaration of a context token,
+         * on the condition that one exists and that the context implies completion should be given.
+         */
+        function tryGetClassLikeCompletionContainer(contextToken: Node): ClassLikeDeclaration {
+            if (contextToken) {
+                switch (contextToken.kind) {
+                    case SyntaxKind.OpenBraceToken:  // class c { |
```
Comment (mhegazy) [inline src/services/completions.ts]: index signature, and computed property name:
```ts 
class c {
   |  ["foo"]  // <- i would expect the keywords at least
}
```

File: src/services/completions.ts [hunk 4]
```diff
@@ -914,6 +924,52 @@ namespace ts.Completions {
         }
 
         /**
+         * Aggregates relevant symbols for completion in class declaration
+         * Relevant symbols are stored in the captured 'symbols' variable.
+         *
+         * @returns true if 'symbols' was successfully populated; false otherwise.
+         */
+        function tryGetClassLikeCompletionSymbols(classLikeDeclaration: ClassLikeDeclaration): boolean {
```
Comment (mhegazy) [inline src/services/completions.ts]: this function only returns `true`, it should be `getClassLikeCompletionSymbols`

File: src/services/completions.ts [hunk 5]
```diff
@@ -914,6 +924,52 @@ namespace ts.Completions {
...
+
+                // No member list for private methods
+                if (!(classElementModifierFlags & ModifierFlags.Private)) {
+                    const baseType = typeChecker.getTypeAtLocation(baseTypeNode);
+                    const typeToGetPropertiesFrom = (classElementModifierFlags & ModifierFlags.Static) ?
+                        typeChecker.getTypeOfSymbolAtLocation(baseType.symbol, classLikeDeclaration) :
+                        baseType;
+
+                    // List of property symbols of base type that are not private
+                    symbols = filter(typeChecker.getPropertiesOfType(typeToGetPropertiesFrom),
```
Comment (mhegazy) [inline src/services/completions.ts]: we should filter ones that are already implemented in the current class.

File: src/services/completions.ts [hunk 6]
```diff
@@ -914,6 +924,52 @@ namespace ts.Completions {
...
+                            break;
+                        case "static":
+                            classElementModifierFlags = classElementModifierFlags | ModifierFlags.Static;
+                            break;
+                    }
+                }
+
+                // No member list for private methods
+                if (!(classElementModifierFlags & ModifierFlags.Private)) {
+                    const baseType = typeChecker.getTypeAtLocation(baseTypeNode);
```
Comment (mhegazy) [inline src/services/completions.ts]: what about interfaces?

File: src/services/completions.ts (hunk 5 above)
Comment (sheetalkamat) [inline src/services/completions.ts]: @DanielRosenwasser commented on the issue saying not to do that. Reference https://github.com/Microsoft/TypeScript/issues/7158#issuecomment-188991281

File: src/services/completions.ts (hunk 6 above)
Comment (sheetalkamat) [inline src/services/completions.ts]: Do we want to show the Implements clause members too? The issue didn't clearly mention it.

File: src/services/completions.ts (hunk 4 above)
Comment (sheetalkamat) [inline src/services/completions.ts]: Was just following the pattern, but I will change it

File: src/services/completions.ts (hunk 5 above)
Comment (mhegazy) [inline src/services/completions.ts]: we do filtering in object literal completions, ppl seem to appreciate that.

File: src/services/completions.ts (hunk 6 above)
Comment (mhegazy) [inline src/services/completions.ts]: well.. probably, do not see why not. like it is likely that it is a member you are implementing or extending..

**Pull request merged**